### PR TITLE
Move network interface authz to the data store

### PR DESCRIPTION
**Pull request by bnaecker**
- Adds a module-private function for actually inserting the database
  record, after performing authz checks. This is used in the
  publicly-available method and in tests.
- Adds authz objects to the
  `DataStore::instance_create_network_interface` method and does authz
  checks inside them.
- Reorders the instance-creation saga. This moves the instance DB record
  creation before the NIC creation, since the latter can only be
  attached to an existing instance record. This also allows uniform
  authz checks inside the `DataStore` method, which wouldn't be possible
  if the instance record were not yet in the database. Note that this
  also requires a small change to the data the instance-record-creation
  saga node serializes. It previously contained the NICs, but these are
  no longer available at that time. Instead, the NICs are deserialized
  from the saga node that creates them and used to instantiate the
  instance runtime object only inside the `sic_instance_ensure` saga
  node.
- Moves authz check for listing NICs for an instance into `DataStore`
  method
- Moves authz check for fetching a single NIC for an instance into the
  `DataStore` method
- Adds the `network_interface_fetch` method, for returning an authz
  interface and the database record, after checking read access. This
  uses a `*_noauthz` method as well, both of which are analogous to the
  other similarly-named methods. Note there's no lookup by ID or path at
  this point, since they're not really needed yet.
- Moves the check for deleting an interface into the `DataStore` method.
- Changes how deletion of a previously-deleted NIC works. We used to
  return a success, but we now return a not-found error, in line with
  the rest of the API.

### CODE REVIEW

File: nexus/src/db/datastore.rs [hunk 1]
```diff
@@ -1676,8 +1676,29 @@ impl DataStore {
...
+            .await
+            .map_err(NetworkInterfaceError::External)?;
+        opctx
+            .authorize(authz::Action::CreateChild, authz_subnet)
+            .await
+            .map_err(NetworkInterfaceError::External)?;
+        self.instance_create_network_interface_raw(&opctx, interface).await
+    }
+
+    pub(super) async fn instance_create_network_interface_raw(
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/datastore.rs:1699]: I tried that, but those records must actually exist in the database if I'm understanding correctly. Or at least, my attempts failed with not-found errors looking up those mocked resources. I can try again, if that's markedly better than this approach?

File: nexus/src/db/datastore.rs [hunk 2]
```diff
@@ -1720,109 +1745,126 @@ impl DataStore {
...
+    pub async fn instance_delete_network_interface(
         &self,
-        subnet_id: &Uuid,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
         interface_name: &Name,
-    ) -> LookupResult<db::model::NetworkInterface> {
-        use db::schema::network_interface::dsl;
+    ) -> DeleteResult {
+        opctx.authorize(authz::Action::Modify, authz_instance).await?;
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/datastore.rs:1810]: That all sounds good. I think part of the lingering confusion (for me) is that NICs are networking abstractions, but they're attached to a specific instance. It now seems like their parent is the instance, but, at least in the past, the VPC Subnet seemed more appropriate.

In any case, I'll update as you recommend. I agree that having the more complicated behaviors described in the Oso policy file is more appropriate.

File: nexus/src/db/datastore.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/db/datastore.rs:1699]: It depends on how you cons them up.  If you just use `FLEET.organization(...).project(...)`, I would not expect you to get an error.  If you use `vpc_lookup_...`, then yes, that won't work.

File: nexus/src/db/datastore.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/db/datastore.rs:1810]: Cool.  Thanks!  The rest of this looks great.

File: nexus/src/db/datastore.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/datastore.rs:1699]: Hmm. I did the former. I must have either misunderstood the error, or not done it correctly. I'll try it again.

[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/db/datastore.rs:1699]: No worries.  I might be misremembering the code path.  This is not important to do now.

File: nexus/src/db/datastore.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/datastore.rs:1810]: Just to be clear, I'm going to keep the CreateChild on the VPC Subnet. I'll change this to require CreateChild on the Instance when creating a new NIC, and Delete on the NIC itself when deleting. I'm not sure if the intention is for the resource graph to be a tree or not, but NICs do seem like a special case where we're creating resources "under" both VPC Subnets and Instances. Does that sound OK, or would we like to have a single permission check for each of these?

File: nexus/src/db/datastore.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/datastore.rs:1699]: Excellent, will keep as-is!

File: nexus/src/db/datastore.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/db/datastore.rs:1810]: I think that's fine for this case.

**Pull request merged**